rtc/pcf85363a.c

```diff
@@ -0,0 +1,371 @@
+#include <posix/utils.h>
+
+#include <i2c.h>
+#include <rtc.h>
+
+/* NOTEs:
+ *  - PCF85363A auto-increments register address post-read, so we can read multiple consecutive registers with single read
+ *  - we configure / use TSR3 register to hold Last Timestamp the RTC switched to battery operation (last "POWERFAIL" event)
+ *  - we configure / use TSR2 register to hold First Timestamp the RTC switched to battery operation (real "POWERFAIL" event?)
+ *  - we use battery-backed generic RTC RAM memory to check if battery supply has failed in the meantime
+ *  - RTC alarm generation with interrupt pin is not used as we're not planning on putting main CPU to sleep
+ */
+
+static const uint8_t DEV_ADDR = 0x51u;             /* PCF85363A */
+static const uint8_t RTC_TIME_REG_ADDR = 0x01;     /* RTC time: seconds register */
+static const uint8_t RTC_TSR2_REG_ADDR = 0x17;     /* RTC event (2): seconds register */
+static const uint8_t RTC_TSR3_REG_ADDR = 0x1d;     /* RTC event (3): seconds register */
+static const uint8_t RTC_TSR_MODE_REG_ADDR = 0x23; /* RTC trigger events configuration register */
+static const uint8_t RTC_FLAGS_REG_ADDR = 0x2b;    /* RTC status flags */
+static const uint8_t RTC_RAM_BYTE_REG_ADDR = 0x2c; /* single byte of battery-backed RAM */
+
+/* clang-format off */
+enum { dev_id_rtc, dev_id_first_powerfail, dev_id_last_powerfail, dev_id_batt_powerfail };
+/* clang-format on */
+
+static struct {
+	unsigned int battery_failed;
+} common;
+
+
+static unsigned int from_bcd(uint8_t byte)
+{
+	return (byte >> 4) * 10 + (byte & 0xf);
+}
+
+
+static uint8_t to_bcd(uint8_t val)
+{
+	return ((val / 10) << 4) | (val % 10);
+}
+
+
+static int rtc_get_time(struct rtc_time *rtc_time, uint8_t reg_addr, int no_weekday)
+{
+	uint8_t timestamp[7] = { 0 };
+	uint32_t data_len = sizeof(timestamp) - (no_weekday ? 1 : 0);
+
+	memset(rtc_time, 0, sizeof(*rtc_time));
+	if (i2c_regRead(DEV_ADDR, reg_addr, timestamp, data_len) < 0)
+		return -EIO;
+
+	rtc_time->tm_sec = from_bcd(timestamp[0] & 0x7f);
+	rtc_time->tm_min = from_bcd(timestamp[1] & 0x7f);
+	rtc_time->tm_hour = from_bcd(timestamp[2]);
+
+
+	if (no_weekday) {
+		/* no weekdays in alarm / event time registers */
+		rtc_time->tm_mday = from_bcd(timestamp[3]);       /* 1-31 */
+		rtc_time->tm_mon = from_bcd(timestamp[4]) - 1;    /* 0-11; RTC holds 1-12 */
+		rtc_time->tm_year = 100 + from_bcd(timestamp[5]); /* year since 1900, RTC holds only last 2 digids - assume 20XX */
+	}
+	else {
+		rtc_time->tm_mday = from_bcd(timestamp[3]);       /* 1-31 */
+		rtc_time->tm_wday = from_bcd(timestamp[4]);       /* weekday 0-6; 0: sunday (RTC has the same semantics) */
+		rtc_time->tm_mon = from_bcd(timestamp[5]) - 1;    /* 0-11; RTC holds 1-12 */
+		rtc_time->tm_year = 100 + from_bcd(timestamp[6]); /* year since 1900, RTC holds only last 2 digids - assume 20XX */
+	}
+
+
+	rtc_time->tm_yday = 0;  /* FIXME?: yday not stored in RTC */
+	rtc_time->tm_isdst = 0; /* RTC holds UTC time, no DST */
```

Comment:
Earlier it is zeroing the entire structure.

Reply:
In here the comments are important, not the code. Leaving as-is.